src/pubsub/index.js

```diff
@@ -0,0 +1,689 @@
+    }
+
+    // Check the message signature if present
+    if (message.signature && !(await verifySignature(message))) {
+      throw errcode(new Error('Invalid message signature'), codes.ERR_INVALID_SIGNATURE)
+    }
+
+    for (const topic of message.topicIDs) {
+      const validatorFn = this.topicValidators.get(topic)
+      if (!validatorFn) {
+        continue
+      }
+      await validatorFn(topic, message)
+    }
+  }
+
+  /**
+   * Normalizes the message and signs it, if signing is enabled.
+   * Should be used by the routers to create the message to send.
+   * @private
+   * @param {Message} message
+   * @returns {Promise<Message>}
+   */
+  _buildMessage (message) {
+    const msg = utils.normalizeOutRpcMessage(message)
+    if (this.signMessages) {
+      return signMessage(this.peerId, msg)
+    } else {
+      return message
+    }
+  }
+
+  // API METHODS
+
+  /**
+   * Get a list of the peer-ids that are subscribed to one topic.
+   * @param {string} topic
+   * @returns {Array<string>}
+   */
+  getSubscribers (topic) {
+    if (!this.started) {
+      throw errcode(new Error('not started yet'), 'ERR_NOT_STARTED_YET')
+    }
+
+    if (!topic || typeof topic !== 'string') {
+      throw errcode(new Error('a string topic must be provided'), 'ERR_NOT_VALID_TOPIC')
+    }
+
+    const peersInTopic = this.topics.get(topic)
+    if (!peersInTopic) {
+      return []
+    }
+    return Array.from(peersInTopic)
+  }
+
+  /**
+   * Publishes messages to all subscribed peers
+   * @override
+   * @param {Array<string>|string} topics
+   * @param {Buffer} message
+   * @returns {Promise<void>}
+   */
+  async publish (topics, message) {
+    if (!this.started) {
+      throw new Error('Pubsub has not started')
+    }
+
+    topics = utils.ensureArray(topics)
+    this.log('publish', topics, message)
+
+    const from = this.peerId.toB58String()
+    const msgObject = {
+      receivedFrom: from,
+      from: from,
+      data: message,
+      seqno: utils.randomSeqno(),
+      topicIDs: topics
+    }
+
+    // Emit to self if I'm interested and emitSelf enabled
+    this.emitSelf && this._emitMessage(msgObject)
+
+    // send to all the other peers
+    await this._publish(msgObject)
+  }
+
+  /**
+   * Overriding the implementation of publish should handle the appropriate algorithms for the publish/subscriber implementation.
+   * For example, a Floodsub implementation might simply publish each message to each topic for every peer
+   * @abstract
+   * @param {InMessage} message
+   * @returns {Promise<void>}
+   *
+   */
+  _publish (message) {
+    throw errcode(new Error('publish must be implemented by the subclass'), 'ERR_NOT_IMPLEMENTED')
+  }
+
+  /**
+   * Subscribes to topics.
+   * @abstract
+   * @param {Array<string>|string} topics
+   * @param {function} [handler]
+   * @returns {void}
+   */
+  subscribe (topics, handler) {
```

Comment:
Previously, we had this broken into two levels `subscribe` and `_subscribe`.
Perhaps this can be better named, eg: `subscribe` and `_newSubscription` (or some other).
But the outer layer, `subscribe` normalized input and removed existing subscriptions, so that `_newSubscription` was able to be extended in a way that didn't need to renormalize + remove existing subscriptions, but rather acted on known new subscriptions.
Eg: In gossipsub, we want to call `join` only on new subscriptions, so overriding just using this method becomes somewhat more cumbersome, than overriding a lower level method for "subscribing to new, previously unsubscribed topic".

Reply:
I went with not decoupling it because all these operations seem mandatory to do. For the gossipsub example, I changed it to:

```js
subscribe (topics: string[], handler: (msg: any) => void): void {
    super.subscribe(topics, handler)
    this.join(topics)
  }
```

Any reason you foresee to not simplify this?

Reply:
I think this is fine, especially considering we'll only be operating on a single topic.